include/sparrow/utils/metadata.hpp

```diff
@@ -0,0 +1,196 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <concepts>
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <numeric>
+#include <ranges>
+#include <string_view>
+#include <utility>
+
+#include "sparrow/config/config.hpp"
+#include "sparrow/utils/contracts.hpp"
+#include "sparrow/utils/pair.hpp"
+
+
+#if defined(__cpp_lib_format)
+#    include <format>
+#    include <ostream>
+#endif
+
+namespace sparrow
+{
+    using metadata_key = std::string_view;
+    using metadata_value = std::string_view;
+    using metadata_pair = std::pair<metadata_key, metadata_value>;
+
+    // Helper function to extract an int32 from a char buffer
+    SPARROW_API int32_t extract_int32(const char*& ptr);
+
+    class key_value_view;
+
+    class key_value_view_iterator
+    {
+    public:
+
+        using iterator_category = std::input_iterator_tag;
+        using value_type = metadata_pair;
+        using difference_type = std::ptrdiff_t;
+        using pointer = value_type*;
+        using reference = value_type&;
+
+        SPARROW_API key_value_view_iterator(const key_value_view& parent, int32_t index);
+
+        SPARROW_API value_type operator*() const;
+
+        SPARROW_API key_value_view_iterator& operator++();
+
+        friend bool operator==(const key_value_view_iterator& lhs, const key_value_view_iterator& rhs)
+        {
+            return lhs.m_index == rhs.m_index;
+        }
+
+        friend bool operator!=(const key_value_view_iterator& lhs, const key_value_view_iterator& rhs)
+        {
+            return !(lhs == rhs);
+        }
+
+    private:
+
+        SPARROW_API std::string_view extract_string_view();
+        SPARROW_API void extract_key_value();
+
+        const key_value_view& m_parent;
+        int32_t m_index;
+        const char* m_current;
+        std::string_view m_key;
+        std::string_view m_value;
+    };
+
+    // Custom view to lazily extract key/value pairs from the buffer
+    class key_value_view : public std::ranges::view_interface<key_value_view>
```

Comment:
I guess this is the reason for dropping GCC12 support? If so, we can try to reintroduce it after this PR is merged.